### ISSUE TITLE
kcov segfaults without verify, with verify it reports 0 invalid breakpoints

### ISSUE THREAD

**Issue by shahn** [SimonKagstrom/kcov#198]
Trying to get coverage for a Rust program, I'm running kcov on it. Without --verify I get a segfault:

    kcov: Process exited with signal 11 (SIGSEGV) at 0x5555555ea18a

but with verify I get

    kcov: 0 invalid breakpoints skipped in target/debug/gen_diff-a0f172fceb51b63d
    kcov: 0 invalid breakpoints skipped in /lib/x86_64-linux-gnu/libdl.so.2
    kcov: 0 invalid breakpoints skipped in /lib/x86_64-linux-gnu/librt.so.1
    kcov: 0 invalid breakpoints skipped in /lib/x86_64-linux-gnu/libpthread.so.0
    kcov: 0 invalid breakpoints skipped in /lib/x86_64-linux-gnu/libc.so.6
    kcov: 0 invalid breakpoints skipped in /lib/x86_64-linux-gnu/libm.so.6

so I'm not sure what's going on. What other kind of stuff could be causing this behaviour? It definitely seems --verify needs to report which problem it actually worked around.

**Comment by SimonKagstrom** [comment]
Yes, this is the same root cause as for Issue #153 and issue #172 and probably a few other bug reports.

However, the number of invalid breakpoints sounds like it's not counting up correctly, or that some file hasn't been reported.

If you run with ELF-debugging,

```
kcov --debug=4 <other args>
```

you should hopefully see something like `kcov: address 12345 is not on an instruction boundary, skipping`. If so, then the counting is definately wrong.

That said, probably the number of skipped breakpoints should probably only be printed as a debug message as well, it's typically not very interesting.

**Comment by shahn** [comment]
I thought it helpful because it might lead to the discovery of where the exact issue lies - it's kind of hard to debug otherwise, if the only thing we have is "dwarf is wrong" :)

This is the debug=4 output:

    Cannot open linux-vdso.so.1
    No debug symbols in target/cov/libkcov_sowrapper.so.
     CRC mismatch for debug link /lib/x86_64-linux-gnu/libdl-2.19.so. Should be 0x703c0b45, is 0xb45adfc1!
    CRC mismatch for debug link /lib/x86_64-linux-gnu/librt-2.19.so. Should be 0xe9b549b6, is 0x0a845a34!
    CRC mismatch for debug link /lib/x86_64-linux-gnu/libpthread-2.19.so. Should be 0x27ac05c0, is 0x8e1fae0f!
    CRC mismatch for debug link /lib/x86_64-linux-gnu/libgcc_s.so.1. Should be 0x15fbad25, is 0xe1528828!
    No debug symbols in /lib/x86_64-linux-gnu/libgcc_s.so.1.
    CRC mismatch for debug link /lib/x86_64-linux-gnu/libc-2.19.so. Should be 0x8318c513, is 0xcd8269aa!
    No debug symbols in /lib64/ld-linux-x86-64.so.2.
    CRC mismatch for debug link /lib/x86_64-linux-gnu/libm-2.19.so. Should be 0xcd531c90, is 0xb34e2f42!

**Comment by shahn** [comment]
Ah, and running with --verify --debug=4, I get a ton of address is not an instruction boundary messages.

**Comment by SimonKagstrom** [comment]
Yes, as explained elsewhere the DWARF information contains an entry for, say, `file.c:123 -> 0x8049df5`, but instructions look something like this:

```
 8049df1:	55                   	push   %ebp
 8049df2:	89 e5                	mov    %esp,%ebp
 8049df4:	8b 45 08             	mov    0x8(%ebp),%eax  # <--- ouch!
 8049df7:	8b 00                	mov    (%eax),%eax
```

so setting a breakpoint (instruction 0xcc) on 0x8049df5 will effectively destroy the instruction. So this is definately a "dwarf is wrong" case :-).


I'll keep the bug report open to track the broken "skipped counts".

**Comment by SimonKagstrom** [comment]
The skipped counts has been fixed with 9c14006.
